File: engine/tools/wearability-bench.mjs
// Wearability benchmark driver. Reads the latest benchmark-58 results JSON (the
// vision-read specs for the 54 real product photos), maps each garment spec to
// engine params the SAME way web/js/create.js does for the opening-relevant
// fields (closure -> placket/tie, backDetail -> back opening), emits a TSV, and
// pipes it through the compiled wearability-bench CLI to count how many specs the
// engine currently drafts into an UNWEARABLE pattern.
//
// Usage: node engine/tools/wearability-bench.mjs [results.json] [wearability-bench-binary]
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { execFileSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const root = join(dirname(fileURLToPath(import.meta.url)), '..', '..');

// --- locate the newest results-*.json (the labelled vision reads) --------------
let resultsPath = process.argv[2];
if (!resultsPath) {
  const dir = join(root, 'benchmark-58');
  const files = existsSync(dir)
    ? readdirSync(dir).filter((f) => /^results-\d{4}-\d\d-\d\d\.json$/.test(f)).sort()
    : [];
  if (!files.length) { console.error('no benchmark-58/results-*.json found (data is local/gitignored)'); process.exit(2); }
  resultsPath = join(dir, files[files.length - 1]);
}
const bin = process.argv[3] || join(root, 'engine', 'build', 'wearability-bench');
if (!existsSync(bin)) { console.error('build the CLI first: cmake --build engine/build (target wearability-bench)'); process.exit(2); }

console.error('results:', resultsPath);
const results = JSON.parse(readFileSync(resultsPath, 'utf8'));

// --- opening mapping, mirroring create.js pickTiePlacement / pickBackOpening ----
const TIE = { none: 0, backWaist: 1, backWaistBow: 2, frontNeckBow: 3, tieBack: 4, cuffTies: 5 };
const BACK = { none: 0, round: 1, lowV: 2, square: 3, keyhole: 4 };

function pickTiePlacement(seen) {
  const oov = Array.isArray(seen.outOfVocab) ? seen.outOfVocab.join(' | ') : '';
  const isDrawstring = (s) => /drawstring|gathered|shirr|smock/.test((s || '').toLowerCase());
  if (seen.backDetail === 'tieBack') return 'tieBack';
  const c = seen.closure;
  if (c && (c.type === 'ties')) {
    const loc = (c.location || '').toLowerCase();
    if (loc.includes('neck') && isDrawstring(oov)) return 'none';
    if (loc.includes('waist')) return 'backWaistBow';
    if (loc.includes('back')) return 'backWaistBow';
    if (loc.includes('neck')) return isDrawstring(oov) ? 'none' : 'frontNeckBow';
    if (loc.includes('front') || loc.includes('center')) return 'frontNeckBow';
    return 'backWaistBow';
  }
  return 'none';
}
function pickBackOpening(seen) {
  const words = [seen.backDetail, Array.isArray(seen.outOfVocab) ? seen.outOfVocab.join(' | ') : '']
    .filter(Boolean).join(' ').toLowerCase();
  const named = /open.?back|back.?cutout|backless|low open back/.test(words) || seen.backDetail === 'openBack';
  if (!named) return null;
  if (/keyhole/.test(words)) return 'keyhole';
  if (/square/.test(words)) return 'square';
  if (/\bv-?\b|low-?v|deep v|v cut|v-cut|plunge/.test(words)) return 'lowV';
  return 'round';
}
function pickFrontPlacket(seen) {
  const c = seen.closure;
  if (c && (c.type === 'buttons' || c.type === 'placket')) {
    const loc = (c.location || '').toLowerCase();
    // create.js: a front/center button closure is drawn as a placket; a back/side
    // closure stays honest.
    if (!loc || loc.includes('front') || loc.includes('center') || loc.includes('cf')) return true;
  }
  return false;
}

// --- normalize the vision garment/neckline/sleeve to engine strings ------------
function garmentOf(s) { return s.garment === 'skirt' ? 'skirt' : s.garment === 'top' ? 'top' : 'dress'; }
function necklineOf(s) {
  const n = (s.neckline || '').toLowerCase();
  if (n.includes('scoop')) return 'scoop';
  if (n.includes('v')) return 'vNeck';
  if (n.includes('square')) return 'square';
  if (n.includes('boat') || n.includes('bateau')) return 'boat';
  if (n.includes('sweetheart') || n.includes('heart')) return 'sweetheart';
  if (n.includes('halter')) return 'halter';
  return 'crew';
}
function sleeveOf(s) {
  const v = (s.sleeveStyle || '').toLowerCase();
  if (v === 'straight' || v === 'set-in' || v === 'fitted') return 'straight';
  if (v === 'balloon' || v === 'puff' || v === 'bishop') return 'balloon';
  return 'none';
}

const rows = [];
for (const [key, entry] of Object.entries(results)) {
  const s = entry.spec;
  if (!s || !s.garment || s.garment === 'other') continue; // rejects/non-garments
  const tie = pickTiePlacement(s);
  const back = pickBackOpening(s);
  const fp = pickFrontPlacket(s);
  const id = (entry.name || key).replace(/\s+/g, '_').slice(0, 28);
  rows.push([
    id, garmentOf(s), necklineOf(s), sleeveOf(s),
    fp ? '1' : '0', String(TIE[tie] ?? 0), String(back ? (BACK[back] ?? 0) : 0),
    s.keyhole ? '1' : '0',
  ].join('\t'));
}

const tsv = rows.join('\n') + '\n';
const out = execFileSync(bin, { input: tsv, encoding: 'utf8' });
process.stdout.write(out);

File: engine/tools/wearability-bench.cpp
// Wearability benchmark harness. Reads spec rows on stdin (one per benchmark
// photo), drafts each on a standard adult body, runs the wearability gate, and
// prints per-row PASS/FAIL + a summary. Lets us count how many of the 54
// benchmark specs the engine currently drafts into an UNWEARABLE pattern.
//
// Input is TSV (tab-separated), one row per garment, fields (all strings/ints):
//   id  garment  neckline  sleeveStyle  frontPlacket  tieClosure  backOpening  keyhole
// frontPlacket/keyhole are 0/1; tieClosure/backOpening are the enum ints.
//
// The mapping from vision language -> these fields is done in the node driver
// (wearability-bench.mjs), mirroring create.js. This CLI is the deterministic
// engine side.
#include <cstdio>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "../src/garment.hpp"
#include "../src/validator.hpp"
#include "../src/wearability.hpp"

using namespace stitchu;

static Neckline necklineFrom(const std::string& s) {
    if (s == "scoop") return Neckline::Scoop;
    if (s == "vNeck") return Neckline::VNeck;
    if (s == "square") return Neckline::Square;
    if (s == "boat") return Neckline::Boat;
    if (s == "sweetheart") return Neckline::Sweetheart;
    if (s == "halter") return Neckline::Halter;
    return Neckline::Crew;
}
static SleeveStyle sleeveStyleFrom(const std::string& s) {
    if (s == "straight") return SleeveStyle::Straight;
    if (s == "balloon") return SleeveStyle::Balloon;
    return SleeveStyle::None;
}
static GarmentType garmentFrom(const std::string& s) {
    if (s == "skirt") return GarmentType::Skirt;
    if (s == "top") return GarmentType::Top;
    return GarmentType::Dress;
}

int main() {
    const BodyMeasurementsSnapshot m{90, 72, 98, 38, 40, 58, 36}; // EU ~38 adult

    int total = 0, fail = 0, headFail = 0, foldFail = 0, edgeFail = 0;
    std::string line;
    std::printf("%-28s %-8s %-10s %-6s  RESULT\n", "id", "garment", "neckline", "opng");
    while (std::getline(std::cin, line)) {
        if (line.empty() || line[0] == '#') continue;
        std::stringstream ss(line);
        std::string id, garment, neckline, sleeveStyle, fp, tie, back, key;
        std::getline(ss, id, '\t');
        std::getline(ss, garment, '\t');
        std::getline(ss, neckline, '\t');
        std::getline(ss, sleeveStyle, '\t');
        std::getline(ss, fp, '\t');
        std::getline(ss, tie, '\t');
        std::getline(ss, back, '\t');
        std::getline(ss, key, '\t');

        GarmentSpec spec;
        spec.garment = garmentFrom(garment);
        spec.neckline = necklineFrom(neckline);
        spec.sleeveStyle = sleeveStyleFrom(sleeveStyle);
        spec.frontPlacket = (fp == "1");
        spec.tieClosure = tie.empty() ? 0 : std::stoi(tie);
        spec.backOpening = back.empty() ? 0 : std::stoi(back);
        spec.keyhole = (key == "1");

        const DraftedPattern d = GarmentDrafter::draft(spec, m);
        const auto wear = Wearability::issues(spec, m, d);
        const double opening = Wearability::finishedNeckOpeningMM(spec, d);

        bool h = false, f = false, e = false;
        for (const auto& i : wear) {
            if (i.detail.find("cannot be put on") != std::string::npos) h = true;
            else if (i.detail.find("closure was dropped") != std::string::npos) f = true;
            else if (i.detail.find("raw, fraying edge") != std::string::npos) e = true;
        }
        total++;
        std::string tag;
        if (h) { headFail++; tag += "HEAD "; }
        if (f) { foldFail++; tag += "FOLD "; }
        if (e) { edgeFail++; tag += "EDGE "; }
        if (!wear.empty()) fail++;
        std::printf("%-28s %-8s %-10s %-6.0f  %s\n",
            id.substr(0, 28).c_str(), garment.c_str(), neckline.c_str(), opening,
            wear.empty() ? "wearable" : ("UNWEARABLE " + tag).c_str());
    }
    std::printf("\nSUMMARY: %d specs, %d UNWEARABLE (head=%d fold=%d edge=%d)\n",
                total, fail, headFail, foldFail, edgeFail);
    return 0;
}
